feat(relationIdentificationColumns): resolve relationIdentificationColumns (BLS-2661)

mcntai opened the pull request:
1. resolve `relationIdentificationColumnName` in `dump` method
2. resolve `relationIdentificationColumnId` based on `relationIdentificationColumnName` in logic triggered by `apply-schema`
3. fix `updateColumn` method to be able to update relative columns

Commit:
fix `updateColumn` method to update relative columns

Commit:
delete dataSize for all columns except STRING and BOOLEAN

vengrov commented on libs/backendless-console-api.js:
```diff
@@ -353,12 +353,14 @@ class Backendless {
 
     async getAppCustomApiKeys(app) {
-        const appSettings = await this.getAppSettings(app.id)
+        if (app.id) {
```
I don't understand this change. How a Backendless app would not have an id?

mcntai replied:
I will roll it back and make a fix for this case in separate pr

vengrov commented on libs/backendless-console-api.js:
```diff
@@ -381,14 +383,13 @@ class Backendless {
 
         if (isRelation(column)) {
-            path += '/relation'
+            path = path.replace(columnName, 'relation')
```
what is this change for?

mcntai replied:
@vengrov this is a fix basically, it did not work for relation columns at all as api is different.

for regular columns: 

`/console/data/tables/{tableName}/columns/{columnName}`

for relation columns: 

`/console/data/tables/{tableName}/columns/relation`

vengrov commented on libs/backendless-console-api.js:
```diff
@@ -451,18 +452,35 @@ class Backendless {
 
-            //we have to preserve the dataSize for the columns
-            if (column.dataSize && ['INT', 'TEXT'].includes(column.dataType)) {
+            //we preserve dataSize for STRING and BOOLEAN only
+            if (column.dataSize && ['DATETIME', 'INT','TEXT', 'DATA_REF', 'JSON',' POINT', 'DOUBLE',].includes(column.dataType)) {
```
1. If we preserve it for string and boolean, would not it be better to check for these two types only rather than checking for any other types?
2. Why would Boolean have dataSize?
3.
```suggestion
            if (column.dataSize && !['STRING', 'FILE_REF'].includes(column.dataType)) {
```

vengrov commented on libs/backendless-console-api.js:
```diff
@@ -477,6 +495,8 @@ class Backendless {
 
         app.tables = app.tables.filter(table => !['orders_dump', 'tmp'].find(prefix => table.name.startsWith(prefix)))
 
+        resolveRelationIdentificationColumns(app.tables)
```
there is a cleanTable method, which has a cleanRelation method, so all iteration is already done for you. just add identification column resolving there

vengrov commented on libs/comparator/tables.js:
```diff
@@ -39,6 +39,13 @@ const buildColumnsMap = table => {
+                    column.metaInfo || (column.metaInfo = {})
+                    column.metaInfo.relationIdentificationColumnId = relation.metaInfo.relationIdentificationColumnId
+
+                    options.push('relationIdentificationColumnId')
```
You are just pushing the same word for comparison instead of specifying the column **name**. 
the comparison must be made by name, not id. The differences should be printed in human-readable format, and only patching the schema should be done using the column id

vengrov commented on libs/comparator/tables.js:
```diff
@@ -127,6 +160,8 @@ const buildAppTablesMap = apps => {
 }
 
 module.exports = apps => {
+    enrichColumnsWithRelationIdentificationData(...apps)
```
please notice the printDifferences method invocation two rows above.
This is the purpose of this module. to compare apps and print differences in human-readable format

mcntai replied:
@vengrov the `buildAppTablesMap` which is exported from this module is reused in table sync logic.

If we want to be able to print `relatedIdentificationColumnName` difference between `schema` and `target app`, we need to resolve `relatedIdentificationColumnName` from `relatedIdentificationColumnId` for the target app.

And for updating the data, we need to resolve `relatedIdentificationColumnId` from `relatedIdentificationColumnName` for the schema.

My point is this module is used not only for printing but for synching as well, that's why I put this logic here. please let me know if it's still wrong.

Commit:
code refactoring

Commit:
- resolve `relationIdentificationColumnName` during dump inside `dump.cleanRelation()`
- enrich `column.options` with `relationIdentificationColumnName` to be able to print differences as well as update columns

Commit:
code refactoring

vengrov commented on libs/comparator/tables.js:
```diff
@@ -39,6 +39,13 @@ const buildColumnsMap = table => {
                 column.unique && (options.push('UQ'))
                 column.required && (options.push('NN'))
 
+                if (relation.metaInfo?.relationIdentificationColumnId) {
```
There should be no **columnId** on this level of the logic.

Once you retrieved the tables/columns information from backendless - immediately replace columnId by columnName.
This will give you the possibility
- to dump schema without any additional logic
- to compare live apps without any additional logic
- to compare json schemas without any additional logic
- all logic will sit in the only place - backendless-console-api

the column id to name transformation should happen in the getAppDataTables->normalizeTable function

the column name to column id transformation should happen in the 
updateColumn method

mcntai replied:
@vengrov  thank you, I did not come to this myself... please check the refactored version. I tested it locally, seems to be working fine

vengrov commented on libs/comparator/tables.js:
```diff
@@ -39,6 +39,13 @@ const buildColumnsMap = table => {
+                    column.metaInfo || (column.metaInfo = {})
+                    column.metaInfo.relationIdentificationColumnId = relation.metaInfo.relationIdentificationColumnId
+
+                    options.push(relation.metaInfo.relationIdentificationColumnName)
```
please attach a screenshot of how the discrepancy will be reported

mcntai replied:
<img width="838" alt="image" src="https://github.com/Carrydmcc/bact/assets/68130217/48ca21e8-5b75-4d72-a9e4-0a5d0fbb64d5">

vengrov commented on libs/comparator/tables.js:
```diff
@@ -106,7 +113,83 @@ const printDifferences = (apps, appTablesMap) => {
+
+    schema.tables.forEach(table => enrichSchemaColumnsWithRelationIdentificationIds(table, targetAppColumnsByNameByTable));
+    targetApp.tables.forEach(table => enrichTargetAppColumnsWithRelationIdentificationNames(table, targetAppRelationsByIdByTable, targetAppColumnsById));
+};
```
all this new code should be completely removed from this module

mcntai replied:
@vengrov I removed everything except the logic that pushes `relationIdentificationColumnName` to the `options`

Commit:
code refactoring

Commit:
Merge branch 'master' into max/resolve-relationIdentificationColumnName
